test/docker-e2e/e2e_test.go

```diff
@@ -171,3 +187,256 @@ func (s *CelestiaTestSuite) WaitForSync(ctx context.Context, statusClient rpccli
+	}
+	currentHeight := currentStatus.SyncInfo.LatestBlockHeight
+	blocksProduced := currentHeight - startHeight
+
+	// Get validator count to calculate minimum required blocks
+	validators, err := rpcClient.Validators(ctx, &currentHeight, nil, nil)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get validators: %w", err)
+	}
+	numValidators := len(validators.Validators)
+	minBlocksRequired := int64(numValidators * defaultBlocksPerValidator)
+
+	if blocksProduced >= minBlocksRequired {
+		s.T().Logf(blockRequirementMetLogFormat,
+			blocksProduced, minBlocksRequired, numValidators, defaultBlocksPerValidator)
+		return currentHeight, nil
+	}
+
+	additionalBlocksNeeded := minBlocksRequired - blocksProduced
+	s.T().Logf(waitingForBlocksLogFormat,
+		additionalBlocksNeeded, blocksProduced, minBlocksRequired, numValidators, defaultBlocksPerValidator)
+
+	if err := wait.ForBlocks(ctx, int(additionalBlocksNeeded), chain); err != nil {
+		return 0, fmt.Errorf("failed to wait for additional blocks: %w", err)
+	}
+
+	finalStatus, err := rpcClient.Status(ctx)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get final status: %w", err)
+	}
+	return finalStatus.SyncInfo.LatestBlockHeight, nil
+}
+
+// fetchValidatorSets retrieves validator sets at both start and end heights
+func (s *CelestiaTestSuite) fetchValidatorSets(
+	ctx context.Context,
+	rpcClient rpcclient.Client,
+	startHeight, endHeight int64,
+) (*coretypes.ResultValidators, *coretypes.ResultValidators, error) {
+	if endHeight <= startHeight {
+		return nil, nil, fmt.Errorf("invalid height range %d to %d", startHeight, endHeight)
+	}
+
+	blocksProduced := endHeight - startHeight
+	s.T().Logf("Checking validator liveness from height %d to %d (%d blocks)", startHeight, endHeight, blocksProduced)
+
+	startValidators, err := rpcClient.Validators(ctx, &startHeight, nil, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("validators query at start height %d: %w", startHeight, err)
+	}
+
+	endValidators, err := rpcClient.Validators(ctx, &endHeight, nil, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("validators query at end height %d: %w", endHeight, err)
+	}
+
+	return startValidators, endValidators, nil
+}
+
+// fetchProposerAddresses gathers proposer addresses from block headers using efficient batching
+func (s *CelestiaTestSuite) fetchProposerAddresses(
+	ctx context.Context,
+	rpcClient rpcclient.Client,
+	startHeight, endHeight int64,
+) ([]string, error) {
+	var (
+		proposers = make(map[string]struct{})
+
+		// Check blocks after startHeight (exclusive) through endHeight (inclusive)
+		minHeight = startHeight + 1
+		maxHeight = endHeight
+	)
+
+	for minHeight <= maxHeight {
+		// BlockchainInfo returns headers in descending order and may limit the range
+		blockchainInfo, err := rpcClient.BlockchainInfo(ctx, minHeight, maxHeight)
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch block headers for range %d-%d: %w", minHeight, maxHeight, err)
```

Comment:
For consistency with other logs in this PR

```suggestion
			return nil, fmt.Errorf("failed to fetch block headers for range %d to %d: %w", minHeight, maxHeight, err)
```